File: python/practicalApplication/bitmap.py
class MyBitMap:
    def __init__(self, size):
        self.size = size
        self.words = [0] * (self.getWordIndex(size-1) + 1)

    # 判断Bitmap某一位的状态
    def getBit(self, bitIndex):
        if bitIndex < 0 or bitIndex > self.size-1:
            print("超过Bitmap的有效范围")
        wordIndex = self.getWordIndex(bitIndex)
        return (self.words[wordIndex] & (1 << bitIndex)) != 0

    # 把Bitmap某一位设置为true
    def setBit(self, bitIndex):
        if bitIndex < 0 or bitIndex > self.size-1:
            print("超出Bitmap的有效范围")
        wordIndex = self.getWordIndex(bitIndex)
        self.words[wordIndex] |= (1 << bitIndex)

    # 定位Bitmap某一位所对应的word
    def getWordIndex(self, bitIndex):
        # 右移6位，相当于除以64
        return bitIndex >> 6


if __name__ == "__main__":
    bitmap = MyBitMap(128)
    bitmap.setBit(126)
    bitmap.setBit(75)
    #bitmap.setBit(130)
    print(bitmap.getBit(126))
    print(bitmap.getBit(78))
    #print(bitmap.getBit(130))

File: C++/practicalApplication/bitmap.cpp
#include<iostream>
#include<vector>
using namespace std;

class Bitmap{
    private:
        vector<long> words;
        int size;
        int getWordIndex(int bitIndex){
            //右移6位，相当于除以64
            return bitIndex >>6;
        }
    public:
        Bitmap(int size){
            this->size=size;
            this->words =vector<long>(getWordIndex(size-1) +1);
        }

        bool getBit(int bitIndex){
            if(bitIndex <0||bitIndex>size-1){
                cout<<"超过Bitmap有效范围！"<<endl;
                return false;
            }
            int wordIndex =getWordIndex(bitIndex);
            return (words[wordIndex] & (1L<<bitIndex)) !=0;
        }
        void setBit(int bitIndex){
            if(bitIndex <0 ||bitIndex>size-1){
                cout<<"超过Bitmap有效范围！"<<endl;
            return;
            }
            int wordIndex = getWordIndex(bitIndex);
            words[wordIndex] |=(1L <<bitIndex);
        }
};

int main(){
    Bitmap bitmap(128);
    bitmap.setBit(126);
    bitmap.setBit(75);
    //bitmap.setBit(-1);
    cout<<bitmap.getBit(126)<<endl;
    cout<<bitmap.getBit(78)<<endl;
    //cout<<bitmap.getBit(129)<<endl;
    return 0;
}